src/lightning/pytorch/strategies/fsdp.py

```diff
@@ -151,6 +165,11 @@ def __init__(
         )
         self.kwargs = kwargs
 
+    def lightning_module_state_dict(self) -> Dict[str, Any]:
+        """Returns model state."""
+        assert self.model is not None
+        return self.model.state_dict()
```

Comment:
Let's do this instead:

```py
if isinstance(model, FullyShardedDataParallel):
    self.model.module.state_dict()
```

Instead of adding another wrapping layer.

Reply:
So, you don't want to add `_FSDPStrategyModuleWrapper`? For some reason, your approach doesn't work, there are only shared layers in state dict. For example, if model contains 2 layers with 5 and 10 parameters and we are using `size_based_auto_wrap_policy` with 7 as `min_num_params`, then state dict will only contain second layer.

I guess this is because `FullyShardedDataParallel.state_dict()` aggregate states from all nodes, but `model.module` is a regular torch module without any additional logic.

If you don't want to add additional wrapping, then we can do something like this:
```python
if isinstance(model, FullyShardedDataParallel):
    raw_state_dict = self.model.state_dict()
    state_dict = OrderedDict({k.removeprefix("_forward_module"): v for k, v in raw_state_dict.items()})
```

Reply:
I would prefer to avoid the extra layer on top, yes. Then you can call self.model.state_dict(). Just a suggestion. But be aware that it will be only a temporary solution. We will want to eventually rework this for implementing checkpointing with local state dict.

Reply:
So, you are okay with just removing this prefix from state dict? `.removeprefix()` is available only in Python 3.9+, is it okay or it is better to write custom function?

I also can implement dumping state dict according to `state_dict_type`, like in torch [documentation](https://pytorch.org/docs/stable/fsdp.html#torch.distributed.fsdp.FullyShardedDataParallel.state_dict_type). We can add argument to `FSDPStrategy` init method and use it in this function.

Reply:
Interesting overrides I saw in the wild: https://github.com/declare-lab/flan-alpaca/blob/97b07b41041acf6606ba253642d07cc7f18eeba2/training.py#L27-L69

Reply:
Yes, I tried to suggest something like this. I'm just curious if we want to support different types of state dicts? If not, and we always need `FULL_STATE`, then the solution is simple and straightforward.

Reply:
@SpirinEgor Yes feel free to just iterate over the keys and remove the `forward_module.` part in front of all keys. 
I'm also fine with the current wrapper but leaning towards the removal of the prefix. 

See also my comment here: https://github.com/Lightning-AI/lightning/issues/16526#issuecomment-1503541511
This PR takes care of the saving of full-state-dict. We will also need to implement the sharded state eventually and take care of loading.